dynamo/preprocessing/gene_selection.py

```diff
@@ -0,0 +1,752 @@
+        keep_filtered: Whether to keep genes that don't pass the filtering in the adata object. Defaults to True.
+        n_top_genes: the number of top genes based on scoring method (specified by sort_by) will be selected as feature
+            genes. Defaults to 2000.
+        sort_by: the sorting methods to be used to select genes. Should be one of the gini index or
+            dispersion of coefficient variation or fano. Defaults to cv_dispersion.
+        exprs_frac_for_gene_exclusion: threshold of fractions for high fraction genes. Defaults to 1.
+        genes_to_exclude: genes that are excluded from evaluation. Defaults to None.
+        SVRs_kwargs: kwargs for `SVRs`. Defaults to {}.
+
+    Raises:
+        NotImplementedError: the 'sort_by' algorithm is invalid/unsupported.
+    """
+
+    filter_bool = (
+        adata.var["pass_basic_filter"]
+        if "pass_basic_filter" in adata.var.columns
+        else np.ones(adata.shape[1], dtype=bool)
+    )
+
+    if adata.shape[1] <= n_top_genes:
+        filter_bool = np.ones(adata.shape[1], dtype=bool)
+    else:
+        if sort_by == "gini":
+            if layer + "_gini" is not adata.var.keys():
+                Gini(adata)
+            valid_table = adata.var[layer + "_gini"][filter_bool]
+            feature_gene_idx = np.argsort(-valid_table)[:n_top_genes]
+            feature_gene_idx = valid_table.index[feature_gene_idx]
+            filter_bool = filter_bool.index.isin(feature_gene_idx)
+        elif sort_by == "cv_dispersion" or sort_by == "fano_dispersion":
+            if not any("velocyto_SVR" in key for key in adata.uns.keys()):
+                select_genes_by_svr(
+                    adata,
+                    layers=layer,
+                    filter_bool=filter_bool,
+                    algorithm=sort_by,
+                    **SVRs_kwargs,
+                )
+            filter_bool = get_svr_filter(adata, layer=layer, n_top_genes=n_top_genes, return_adata=False)
+        else:
+            raise NotImplementedError(f"The algorithm {sort_by} is invalid/unsupported")
+
+    # filter genes by gene expression fraction as well
+    if "frac" not in adata.var.keys():
+        adata.var["frac"], invalid_ids = compute_gene_exp_fraction(X=adata.X, threshold=exprs_frac_for_gene_exclusion)
+
+    genes_to_exclude = (
+        list(adata.var_names[invalid_ids])
+        if genes_to_exclude is None
+        else genes_to_exclude + list(adata.var_names[invalid_ids])
+    )
+    if genes_to_exclude is not None and len(genes_to_exclude) > 0:
+        adata_exclude_genes = adata.var.index.intersection(genes_to_exclude)
+        adata.var.loc[adata_exclude_genes, "use_for_pca"] = False
+
+    if keep_filtered:
+        adata.var["use_for_pca"] = filter_bool
+    else:
+        adata._inplace_subset_var(filter_bool)
+        adata.var["use_for_pca"] = True
+
+    adata.uns["feature_selection"] = sort_by
+
+
+def select_genes_by_svr(
+    adata_ori: AnnData,
```

Comment:
Since this function will not be called by the Preprocessor. Would it be better to rename it to something like `_select_genes_by_svr`? In this way, we can avoid the confusion of calling one `select_genes` function in another `select_genes` function. Same for `select_genes_by_seurat_dispersion`.

Reply:
Good idea! Fixed.